src/sql/src/pure.rs

```diff
@@ -522,70 +539,74 @@ async fn purify_create_sink(
-                key:
-                    Format::Avro(AvroSchema::Csr {
-                        csr_connection: CsrConnectionAvro { connection, .. },
-                    }),
-                value: _,
-            } => {
-                let connection = {
-                    let scx = StatementContext::new(None, &catalog);
-                    let item = scx.get_item_by_resolved_name(&connection.connection)?;
-                    // Get Kafka connection
-                    match item.connection()? {
-                        Connection::Csr(connection) => {
-                            connection.clone().into_inline_connection(&catalog)
-                        }
-                        _ => Err(CsrPurificationError::NotCsrConnection(
-                            scx.catalog.resolve_full_name(item.name()),
-                        ))?,
-                    }
-                };
+        };
 
-                let client = connection
-                    .connect(storage_configuration, InTask::No)
-                    .await
-                    .map_err(|e| CsrPurificationError::ClientError(Arc::new(e)))?;
+        let client = connection
+            .connect(storage_configuration, InTask::No)
+            .await
+            .map_err(|e| CsrPurificationError::ClientError(Arc::new(e)))?;
 
-                client
-                    .list_subjects()
-                    .await
-                    .map_err(|e| CsrPurificationError::ListSubjectsError(Arc::new(e)))?;
-            }
-            FormatSpecifier::Bare(
-                Format::Avro(AvroSchema::InlineSchema { .. })
-                | Format::Bytes
-                | Format::Csv { .. }
-                | Format::Json { .. }
-                | Format::Protobuf(ProtobufSchema::InlineSchema { .. })
-                | Format::Regex(..)
-                | Format::Text,
-            )
-            | FormatSpecifier::KeyValue { .. } => {}
-        }
+        client
+            .list_subjects()
+            .await
+            .map_err(|e| CsrPurificationError::ListSubjectsError(Arc::new(e)))?;
     }
 
+    purify_create_sink_avro_doc_on_options(&catalog, *from.item_id(), format)?;
+
     Ok(PurifiedStatement::PurifiedCreateSink(create_sink_stmt))
 }
 
+/// Runs a function on each format within the format specifier.
+///
+/// The function is provided with a `DocOnSchema` that indicates whether the
+/// format is for the key, value, or both.
+///
+// TODO(benesch): rename `DocOnSchema` to the more general `FormatRestriction`.
+fn for_each_format<'a, F>(format: &'a mut Option<FormatSpecifier<Aug>>, mut f: F)
+where
+    F: FnMut(DocOnSchema, &'a mut Format<Aug>),
+{
+    match format {
+        None => (),
+        Some(FormatSpecifier::Bare(fmt)) => f(DocOnSchema::All, fmt),
+        Some(FormatSpecifier::KeyValue { key, value }) => {
+            f(DocOnSchema::KeyOnly, key);
+            f(DocOnSchema::ValueOnly, value);
+        }
+    }
+}
+
```

Comment:
this is a great abstraction! I think there are likely other places we refactor to use something like this in the source render pipeline when we're decoding key/value formats